### PR TITLE
refactor(compiler-cli): cleanup after TS 4.0 changes

### PR DESCRIPTION
**Pull request by alxhub**
This commit extracts some TS 4.0 sensitive operations into utility functions
to better isolate them from the rest of the code base. This way, support for
3.9-specific functionality is isolated to a single location, and cleanup
after removal of 3.9 support will be much more straightforward.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/transform/src/declaration.ts [hunk 1]
```diff
@@ -275,9 +273,16 @@ export class ReturnTypeTransform implements DtsTransform {
...
+          /* node */ element,
+          /* decorators */ element.decorators,
+          /* modifiers */ element.modifiers,
+          /* asteriskToken */ element.asteriskToken,
+          /* name */ element.name,
+          /* questionToken */ element.questionToken,
+          /* typeParameters */ element.typeParameters,
+          /* parameters */ element.parameters,
+          /* type */ tsReturnType,
+          /* body */ element.body);
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/transform/src/declaration.ts:285]: These comments seem a bit overkill :-) Why did you add them, when we don't tend to do this for other calls.

File: packages/compiler-cli/src/ngtsc/util/src/typescript.ts [hunk 2]
```diff
@@ -143,3 +143,72 @@ export function resolveModuleName(
...
+export function isLiteralNullTypeNode(node: ts.TypeNode): boolean {
+  return node.kind === ts.SyntaxKind.NullKeyword ||
+      (ts.isLiteralTypeNode(node) && node.literal.kind === ts.SyntaxKind.NullKeyword);
+}
+
+/**
+ * Mutate the `pos` and `end` of a `ts.NodeArray`.
+ *
+ * This operation is not strictly legal per the type declaration of `ts.NodeArray`, but is supported
+ * for single nodes via `ts.setSourceMapRange`. Likely this is just an omission from the TS API.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/util/src/typescript.ts:208]: 👍

**Review by petebacondarwin — APPROVED** [review]
I had wondered about doing exactly this, but I got the impression that we would move off 3.9 very soon and Alan was going to immediately follow up with a new refactoring. But I agree it is probably for the best whatever the timeframe.

**Comment by petebacondarwin** [comment]
Oops it actually looks like there are some compilation errors...

File: packages/compiler-cli/src/ngtsc/transform/src/declaration.ts (hunk 1 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/transform/src/declaration.ts:285]: It actually is generally my pattern to annotate the `ts.create*` calls like this. This one is a bit excessive because we're passing through the original values for most of the parameters, but commonly these calls involve a lot of `undefined` values (for things like `asteriskToken`, `questionToken`, etc), so having the names annotated is helpful. I much prefer the way the formatter lays out each parameter on a separate line with the annotations, too.

**Review by alan-agius4 — APPROVED** [review]
Nice cleanup, thanks for this!

File: packages/compiler-cli/src/ngtsc/metadata/src/util.ts [hunk 3]
```diff
@@ -10,7 +10,7 @@ import * as ts from 'typescript';
 
 import {Reference} from '../../imports';
 import {ClassDeclaration, ClassMember, ClassMemberKind, isNamedClassDeclaration, ReflectionHost, reflectTypeEntityToDeclaration} from '../../reflection';
-import {nodeDebugInfo} from '../../util/src/typescript';
+import {getTupleTypeNodeElements, nodeDebugInfo} from '../../util/src/typescript';
```
Comment (alan-agius4) [inline packages/compiler-cli/src/ngtsc/metadata/src/util.ts:13]: The build is failing because this file is not included as input in `//packages/compiler-cli/src/ngtsc/reflection:reflection`
